canonical/traefik-k8s-operator
# Moves private key from stored state to Juju secret

## Issue
Private key stored in stored state.


## Solution
Using Juju secrets for storing the private key.


## Context
Juju 3 has introduced secret management.


## Testing Instructions
1. Bootstrap your Juju `3.1` controller, for it you'll need microk8s with strict confinement.
2. Deploy `traefik`
```bash
juju deploy ./traefik-k8s_ubuntu-20.04-amd64.charm traefik-ingress --trust --resource traefik-image=docker.io/jnsgruk/traefik:2.6.1
```
3. Deploy `tls-operator` for providing dummy self signed certs
```bash
juju deploy tls-certificates-operator \
  --config generate-self-signed-certificates=true \
  --config ca-common-name=dummycommonname
```
4. Relate both apps
```bash
juju relate traefik-k8s tls-certificates-operator
```

5. Make sure the private key secret is stored using Juju secrets
```bash
❯ juju secrets 
ID                    Owner            Rotation  Revision  Last updated
cfnkrncviroh030nrjbg  traefik-ingress  never            1  52 minutes ago  
```

```bash
❯ juju show-secret cfnkrncviroh030nrjbg                    
cfnkrncviroh030nrjbg:
  revision: 1
  owner: traefik-ingress
  label: PRIVATE_KEY
  created: 2023-02-17T09:50:22Z
  updated: 2023-02-17T09:50:22Z
```

6. Additionally we can check as well going into the container
```bash
❯ microk8s.kubectl exec -n traefik-secrets-3 -it traefik-ingress-0 -c traefik -- /bin/bash
root@traefik-ingress-0:/# ls
bin  boot  charm  dev  etc  home  lib  lib32  lib64  libx32  media  mnt  opt  proc  root  run  sbin  srv  sys  tmp  usr  var
root@traefik-ingress-0:/# cd /opt/traefik/juju
root@traefik-ingress-0:/opt/traefik/juju# ls
certificate.cert  certificate.key  certificates.yaml
```

## Release Notes
Traefik's private key was previously stored in `self._stored`, with the recent introduction of Juju secrets it has been moved there and removed from stored state.
